addons/pos_loyalty/models/loyalty_reward.py

```diff
@@ -11,3 +12,20 @@ def _get_discount_product_values(self):
         for vals in res:
             vals.update({'taxes_id': False})
         return res
+
+    def _should_compute_all_discount_product_pos(self, reward):
+        reward_discount_product_domain = reward._get_discount_product_domain()
+        if not reward_discount_product_domain:
+            return False
+        pos_product_product_loaded_fields = set(
+            self.env['pos.session'].POS_PRODUCT_PRODUCT_FIELDS) | {'id'}
```

Comment:
Since '_loader_params_product_product' is overridden to include additional fields, it may not account for these added fields. Moreover, this issue specifically pertains to the PoS, and '_should_compute_all_discount_product_pos' should return false if it's not related to the PoS.

To address both of these problems, consider using context. Within the "get_pos_ui_product_product_by_params" method, you can add the loaded product fields to the context. When trying to access these fields here, check if they exist in the context. If they are not present in the context, you can safely assume that the operation is unrelated to the PoS.

Reply:
nice catch!
Rather than including a new System Parameter, do you think it would be better to use an existing one that looks to load in PoS context, see;
https://github.com/odoo/odoo/blob/a13d04832576cdd269143e8b5a6bd48c3403bac4/addons/point_of_sale/models/pos_session.py#L1599C28-L1599C28

with a `if 'loaded_data' not in self.env.context:`

Reply:
I also have some doubt that this part of the code will be called outside of the PoS context, I'll install sale to see if it breakpoint on it. Let me know if you think of a particular flow that might triggers it outside of the PoS

Reply:
To accommodate the required product fields, I suggest introducing a new context. You can insert the product fields following this line in the [codebase](https://github.com/odoo/odoo/blob/a13d04832576cdd269143e8b5a6bd48c3403bac4/addons/point_of_sale/models/pos_session.py#L2041).

Reply:
the field values are computed before this. As such, the original issue will persist with the change that you suggest.
Here is the stack path (of when the value is computed):
![image](https://github.com/odoo/odoo/assets/60775325/70b66201-47c4-4965-9607-57fd4ba38b34)

I propose to add the parameter to the `load_pos_data` function

Reply:
Apologies for the oversight. The correct location to insert the product fields is at the following [line](https://github.com/odoo/odoo/blob/a13d04832576cdd269143e8b5a6bd48c3403bac4/addons/point_of_sale/models/pos_session.py#L1973).

Reply:
it does not works neither 
Assuming you mean: `self = self.with_context(**params['context'], pos_loading=True)`
the context is indeed added, but it get removed as soon as the function return
![image](https://github.com/odoo/odoo/assets/60775325/86c5c525-4432-4e39-a00c-df6648389596)

![image](https://github.com/odoo/odoo/assets/60775325/cfefabb7-8b53-4e46-a8fa-6e661c50ceb9)

Reply:
@pedrambiria 
For me:
 1. either we set the context in a high level function (`load_pos_date` for instance) in case in the future other functions might also need to check the context to avoid having a ton of context key with the same purpose
 2. either we set it in a closer level method (`_get_pos_ui_loyalty_reward`) and I set a more precise name for the context key like `pos_loyalty_reward_load`

Personally I would prefer the first idea, what do you think ?

Reply:
My goal was to add the loaded fields to the context as there is a possibility that fields may be added without the use of '_get_pos_product_product_fields'.

Reply:
in this "worst" case it will force the python evaluation and we could adapt the code if necessary. It looks that the loyalty are loaded before which in this case there is no way to know the fields in advance (except calling the function for nothing and check the fields loaded ? ).
In the case that someone does override it in the "previous implementation" it's not a big deal as it will in the worst case force to use the Python evaluation